File: Flatten List Iterator.py
"""
# Design-3

## Problem 1: Flatten Nested List Iterator (https://leetcode.com/problems/flatten-nested-list-iterator/)

Given a nested list of integers, implement an iterator to flatten it.

Each element is either an integer, or a list -- whose elements may also be integers or other lists.

Example 1:

Input: [[1,1],2,[1,1]]
Output: [1,1,2,1,1]
Explanation: By calling next repeatedly until hasNext returns false,
             the order of elements returned by next should be: [1,1,2,1,1].
Example 2:

Input: [1,[4,[6]]]
Output: [1,4,6]
Explanation: By calling next repeatedly until hasNext returns false,
             the order of elements returned by next should be: [1,4,6].

"""
#`class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """

from collections import deque
class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        self.q = deque()
        self.flatten_helper(nestedList)

    def flatten_helper(self, nestedList):
        for element in nestedList:

            if element.isInteger():

                self.q.append(element.getInteger())

            else:
                self.flatten_helper(element.getList())

    def next(self) -> int:
        return self.q.popleft()

    def hasNext(self) -> bool:
        return len(self.q) != 0



class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        self.stack = []
        self.stack.append(nestedList)


    def next(self) -> int:
        self.hasNext()
        return self.stack[-1].next().getInteger()


    def hasNext(self) -> bool:
        while len(self.stack) != 0:
            if self.stack[-1].hasNext():
                self.stack.pop()
            else:
                x = self.stack[-1].next()
                if x.isInteger():
                    return self.stack[-1].previous()==x
                self.stack.append(x.getList())
        return False

File: FlattenListIterator.java
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * public interface NestedInteger {
 *
 *     // @return true if this NestedInteger holds a single integer, rather than a nested list.
 *     public boolean isInteger();
 *
 *     // @return the single integer that this NestedInteger holds, if it holds a single integer
 *     // Return null if this NestedInteger holds a nested list
 *     public Integer getInteger();
 *
 *     // @return the nested list that this NestedInteger holds, if it holds a nested list
 *     // Return null if this NestedInteger holds a single integer
 *     public List<NestedInteger> getList();
 * }
 Time Complexity : O(N)
 Space COmplexity : O(N)
 Idea : Store Iterator pointers on stack. if peek element is Integer then return the value from hasNext
 Now if the iterator is list then push it into thw stack u keep on doing this until u find an integer
 this takes care of depply nested list
 at last u have to remove the iterator from the stack if the next pointed by it is null
 this way the stack will eventually get empty indicating there are no more iterators to process. 
 */
public class NestedIterator implements Iterator<Integer> {
    NestedInteger nextElement;
    Stack<Iterator<NestedInteger>> stack;
    public NestedIterator(List<NestedInteger> nestedList) {
        stack = new Stack<>();
        stack.push(nestedList.iterator());
    }

    @Override
    public Integer next() {
        return nextElement.getInteger();
    }

    @Override
    public boolean hasNext() {
        while(!stack.isEmpty()){
            if(!stack.peek().hasNext()){
                stack.pop();
                continue;
            }
            nextElement = stack.peek().next();
            if(nextElement.isInteger()){
                return true;
            }else{
                stack.push(nextElement.getList().iterator());
            }
        }
        return false;
    }
}